lib/std/math/isnormal.zig

```diff
@@ -1,45 +1,48 @@
 const std = @import("../std.zig");
 const math = std.math;
 const expect = std.testing.expect;
-const maxInt = std.math.maxInt;
 
-// Returns whether x has a normalized representation (i.e. integer part of mantissa is 1).
+/// Returns whether x is neither zero, subnormal, infinity, or NaN.
 pub fn isNormal(x: anytype) bool {
     const T = @TypeOf(x);
-    switch (T) {
-        f16 => {
-            const bits = @bitCast(u16, x);
-            return (bits + (1 << 10)) & (maxInt(u16) >> 1) >= (1 << 11);
-        },
-        f32 => {
-            const bits = @bitCast(u32, x);
-            return (bits + (1 << 23)) & (maxInt(u32) >> 1) >= (1 << 24);
-        },
-        f64 => {
-            const bits = @bitCast(u64, x);
-            return (bits + (1 << 52)) & (maxInt(u64) >> 1) >= (1 << 53);
-        },
-        f128 => {
-            const bits = @bitCast(u128, x);
-            return (bits + (1 << 112)) & (maxInt(u128) >> 1) >= (1 << 113);
-        },
-        else => {
-            @compileError("isNormal not implemented for " ++ @typeName(T));
-        },
+    const TBits = std.meta.Int(.unsigned, @bitSizeOf(T));
+    if (@typeInfo(T) != .Float) {
+        @compileError("isNormal not implemented for " ++ @typeName(T));
     }
+
+    const increment_exp = 1 << math.floatMantissaBits(T);
+    const remove_sign = ~@as(TBits, 0) >> 1;
+
+    // Hack: 1 is added to the exponent, and if it overflows to 0 or becomes 1,
+    // then it was all zeroes (subnormal) or all ones (non-finite, like inf/nan).
+    // The sign bit is removed because all ones would overflow into it,
+    // as well as the sign bit in general getting in the way.
+    const float_bits = @bitCast(TBits, x);
+    const result = float_bits +% increment_exp;
+
+    return result & remove_sign >= (increment_exp << 1);
 }
 
 test "math.isNormal" {
-    try expect(!isNormal(math.nan(f16)));
-    try expect(!isNormal(math.nan(f32)));
-    try expect(!isNormal(math.nan(f64)));
-    try expect(!isNormal(math.nan(f128)));
-    try expect(!isNormal(@as(f16, 0)));
-    try expect(!isNormal(@as(f32, 0)));
-    try expect(!isNormal(@as(f64, 0)));
-    try expect(!isNormal(@as(f128, 0)));
-    try expect(isNormal(@as(f16, 1.0)));
-    try expect(isNormal(@as(f32, 1.0)));
-    try expect(isNormal(@as(f64, 1.0)));
-    try expect(isNormal(@as(f128, 1.0)));
+    inline for ([_]type{ f16, f32, f64, f128 }) |T| {
+        const TBits = std.meta.Int(.unsigned, @bitSizeOf(T));
+        const mantissa_bits = math.floatMantissaBits(T);
+
+        // normals
+        try expect(isNormal(@as(T, 1.0)));
```

Comment:
Maybe add the max and min normals, e.g. `f128_max` and `f128_min` (add a function that takes the float type in `math.zig`?).

Reply:
This comment (and a few others) were marked resolved but the code changes don't seem to be present, just checking this wasn't lost through some git operation or anything like that :)

Reply:
I had some unforeseen issues with running the test suite, using a new environment. Threw it at the CI and got some rest. I was gonna reread all of these and have a big summary typed up for the PR, given it worked.